crates/rpc/src/gas_price.rs

```diff
@@ -1,112 +1,98 @@
 use std::sync::Arc;
-use std::time::Duration;
+use std::time::{Duration, Instant};
 
+use primitive_types::U256;
 use starknet_gateway_client::GatewayApi;
+use starknet_gateway_types::reply::MaybePendingBlock;
 
 /// Caching of starknet's gas price with single request at a time refreshing.
 ///
 /// The `gasPrice` is used for `estimate_fee` when user
 /// requests for [`pathfinder_common::BlockId::Latest`] or  [`pathfinder_common::BlockId::Pending`].
 #[derive(Clone)]
 pub struct Cached {
-    inner: Arc<std::sync::Mutex<Inner>>,
+    lock: Arc<tokio::sync::Mutex<()>>,
+    value: Arc<std::sync::Mutex<Option<Value>>>,
     gateway: starknet_gateway_client::Client,
-    stale_limit: Duration,
+    horizon: Duration,
+}
+
+#[derive(Clone)]
+struct Value {
+    gas_price: U256,
+    updated: Instant,
 }
 
 impl Cached {
     pub fn new(gateway: starknet_gateway_client::Client) -> Self {
         Cached {
-            inner: Default::default(),
+            lock: Default::default(),
+            value: Default::default(),
             gateway,
-            stale_limit: Duration::from_secs(60),
+            horizon: Duration::from_secs(60),
         }
     }
 
     /// Returns either a fast fresh value, slower a periodically polled value or fails because
     /// polling has stopped.
-    pub async fn get(&self) -> Option<primitive_types::U256> {
-        let mut rx = {
-            let mut g = self.inner.lock().unwrap_or_else(|e| e.into_inner());
-
-            if let Some((fetched_at, gas_price)) = g.latest.as_ref() {
-                if fetched_at.elapsed() < self.stale_limit {
-                    // fresh
-                    let accepted = *gas_price;
-                    return Some(accepted);
+    pub async fn get(&self) -> Option<U256> {
+        // Make sure only a single call updates the gas price
+        let _lock = self.lock.lock().await;
+
+        match self.value.try_lock() {
+            Ok(guard) => match *guard {
+                Some(Value { gas_price, updated }) if updated.elapsed() < self.horizon => {
+                    tracing::debug!(?gas_price, from=?updated, "Using cached gas price value");
+                    return Some(gas_price);
+                }
+                _ => {
+                    tracing::debug!("Gas price missing or expired");
                 }
+            },
+            Err(reason) => {
+                tracing::debug!(%reason, "Failed to lock gas price mutex");
```

Comment:
Given that we now have the outer lock -- how does this locking failure happen?

Reply:
Or to rephrase this: wouldn't having a single async mutex around the actual data and `lock()`-ing that be enough?

Having these two mutexes seems excessive to me.